New file for creating synthetic data to add more training data to the tesseract.

JoshuaJAnderson opened the pull request:
# Description

This file will be used to create synthetic data for tesseract to have more data to train on.

[Trello Card](https://trello.com/c/nyFSsSFf/847-as-a-developer-i-want-to-increase-the-accuracy-of-the-tesseract-model-by-adding-new-data-to-be-trained)

[Loom Video](https://www.loom.com/share/c046c5e579144e2fa693af0f2a7d6636)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Change Status

- [ ] Complete, tested, ready to review and merge
- [x] Complete, but not tested (may need new tests)
- [ ] Incomplete/work-in-progress, PR is for discussion/feedback

## Has This Been Tested

- [x] Yes
- [ ] No
- [ ] Not necessary

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] There are no merge conflicts

JoshuaJAnderson committed:
New file for creating synthetic data to add more training data to the tesseract.

gianpaolocombatti reviewed (changes requested):
Hey Josh - Looked at the video and the generator looks awesome, great work! Unfortunately, the data generator in this commit doesn't have any lines of code. Not sure if you uploaded from a different branch. Can you try uploading it again so I can review the actual code? I tried fetching the branch but no luck from that front either.

JoshuaJAnderson committed:
File didn't commit with code inside.

nlittlepoole reviewed (changes requested):
@MicroCapacitor I love what you've got started here but I think what would be most helpful here is to ship this as an experiment so that others can reproduce it. The following is how I would approach doing that

1. create your own experiment folder in `structured_experiments` and name it something like `2022.02.17.synthetic_generation`
2. Copy the code I have in `2022.02.01.test` to your new folder
3. Move `synthetic_data_generator.py` into the folder
4. Modify `synthetic_data_generator.py` so that it generates more than one image (let's say 100) using different strings. Also modify it to save the `.png` and `gt.txt` files to a directory [specified via command line argumet](https://stackoverflow.com/questions/1009860/how-to-read-process-command-line-arguments)
5. Modify the Dockerfile to copy `synthetic_data_generator.py` into the image
6. Modify `experiment.sh` in your folder.  The following should replace the `setup_ground_truth.py` logic. First you want it to create a folder called `synthetic-ground-truth` in `/train/tesstrain/data/`. Then you want to run `synthetic_data_generator.py`, passing it `/train/tesstrain/data/synthetic-ground-truth` as an argument (support added in #4). This will create synthetic images in that folder . Lastly change the `MODEL_NAME` in the make command to use `synthetic` so that Tesseract uses `synthetic-ground-truth` instead of `kaggle-ground-truth`.
7. Run the docker commands in the `experiment.MD` to confirm its working
8. Update the `experiment.MD` with documentation on your experiment

bofori20 reviewed (dismissed):
Great work done. We should be able to get in synthetic data to improve the model.

JoshuaJAnderson committed:
-Changed location of synthetic_data_generator.py to its own testing folder.
-Added folder that contains the images of each character.
-Added a way to create simple strings that can be expanded on.

JoshuaJAnderson committed:
Deleted a commented out code

JoshuaJAnderson committed:
Edited comments and .MD file

nlittlepoole reviewed (changes requested):
@MicroCapacitor Ok so a few more items:

1. Do you only have one image cut out per image? regardless, you should be adding these to the repository with [git-lfs](https://git-lfs.github.com/). We don't want to pollute the repository with big image files.
2. A lot more fleshed out of an experiment.MD file. This is what future cohorts are going to use to really understand the value of what you have done here. Some questions I'd love to see answered are what is the approach for generating new images and where do you find it in the code? (methodology) What are the theoretical limitations of this approach? What were the training results for tesseract? What can we conclude about using synthetic data from those test results? is it viable? are there limitations? This is the science part of data science that we don't want to lose.

JoshuaJAnderson committed:
Added a mkdir to "experiment.sh" so if there is no folder it will be created.
Edited a for loop to look better.
Deleted "setup_ground_truth.py" since it doesn't apply to my experiments.
Slight edit to "experiment.MD" will edit more later on.

JoshuaJAnderson committed:
Removed a line of code from Dockerfile creating an error from previously removed file.

JoshuaJAnderson committed:
Increased permutations of strings to reach almost 22,000 different variations of a 6-word sentence. Changed character image files due to corruption of previous.

JoshuaJAnderson committed:
Removed the folder character_images from .gitignore.

JoshuaJAnderson committed:
Added character images via lfs.

nlittlepoole reviewed (dismissed):
:shipit:

jinjahninjah reviewed (dismissed):
Awesome work Joshua! :shipit:

jinjahninjah committed:
Merge branch 'main' into feature/synthetic_data

nlittlepoole reviewed (approved):
👍
